### PR TITLE
feat: adl add — incremental repo onboarding with scaffolding

### LINKED ISSUE #25: feat: adl add — incremental repo onboarding with scaffolding
Part of #23

## Context

**User Story:** As a user with a working `~/.adl/config.yaml`, I want to run `adl add ~/projects/my-app` to onboard a repository so the daemon starts processing its issues automatically.

Currently, adding a repo means manually editing `config.yaml` to append a `repos:` entry with the correct `path`, `project_number`, `owner`, and `columns` mapping. The user must know the GitHub Projects V2 project number and column names, which requires GraphQL queries or navigating the GitHub UI.

## Complexity Hint

**Subsystems touched:** CLI (`cli.py`), config file (read + append), GitHub API (`gh` CLI)
**Independent stories:** 1
**Structural changes needed:** No

## Proposed Solution

Add an `adl add [path]` command that:

1. **Resolve repo path**
   - If `path` is given, use it; otherwise use current directory
   - Verify it's a git repo with a GitHub remote

2. **Auto-detect GitHub info**
   - Run `gh repo view --json nameWithOwner,owner` to get owner/repo
   - Run `gh project list --owner <owner>` to list Projects V2
   - Let user pick which project (or enter project number)

3. **Pick columns**
   - Query the selected project for its columns/fields
   - Let user map: source column ("Ready for Dev"), in_progress, done
   - Offer defaults if standard column names are detected

4. **Scaffold agents + workflows**
   - Copy bundled agent markdown files from package into `<repo>/agents/`
   - Copy bundled workflow YAMLs into `<repo>/workflows/`
   - Skip files that already exist (don't overwrite customizations)
   - Print what was scaffolded

5. **Append to config**
   - Read existing `~/.adl/config.yaml`
   - Append new entry to `repos:` list
   - Write back, preserving existing content
   - Print: "Added <owner/repo>. Run `adl run` to start processing."

## Acceptance Criteria

- [ ] VERIFY: `adl add ~/projects/my-app` detects the GitHub repo automatically
- [ ] VERIFY: `adl add` (no path) uses the current directory
- [ ] VERIFY: Wizard lists available GitHub Projects V2 and lets user pick one
- [ ] VERIFY: Wizard queries project columns and lets user map source/in_progress/done
- [ ] VERIFY: Bundled agent files are copied to `<repo>/agents/` (skipping existing)
- [ ] VERIFY: Bundled workflow files are copied to `<repo>/workflows/` (skipping existing)
- [ ] VERIFY: New repo entry is appended to `~/.adl/config.yaml`
- [ ] VERIFY: Running `adl add` on an already-added repo warns and skips
- [ ] VERIFY: `adl validate` passes after `adl add`
- [ ] VERIFY: All existing tests pass

## Edge Cases

- Repo has no GitHub Projects V2 → error with instructions to create one
- Repo already in config → warn "already configured", offer to reconfigure
- No `~/.adl/config.yaml` exists → error "run `adl init` first"
- Non-standard column names → let user type custom names
- Package agents/workflows must be discoverable at runtime via `importlib.resources` or `__file__`-relative paths

### Files to Modify

- `src/auto_dev_loop/cli.py` — add `add` command

### Files to Create

- `src/auto_dev_loop/add_repo.py` — repo onboarding logic (GitHub detection, column mapping, scaffolding, config append)

### Approach

1. Create `src/auto_dev_loop/add_repo.py`
2. Use `subprocess.run(["gh", ...])` to call GitHub CLI for project detection and column listing
3. Use `typer.prompt()` for interactive selection
4. Use `importlib.resources` to locate bundled `agents/` and `workflows/` files within the package
5. Copy files with `shutil.copy2`, skipping existing
6. Read config YAML, append to `repos` list, write back with `pyyaml`
7. Add `@app.command() def add()` to `cli.py`

### Patterns to Follow

- Follow existing CLI patterns in `cli.py`
- `RepoConfig` structure in `models.py` defines the expected shape
- GitHub CLI (`gh`) is already used in `poller.py` — consistent approach

### Constraints

- Must require `gh` CLI to be installed and authenticated
- Must not modify existing repo entries when appending
- Scaffolded files must be the same ones shipped in the package's `agents/` and `workflows/` dirs

### In Scope
- `adl add [path]` interactive command
- GitHub project detection and column mapping
- Agent/workflow scaffolding
- Config file append

### Out of Scope
- Repo removal (`adl remove` — future work)
- Config migration
- Validating that the GitHub project actually has issues

## Dependencies

- Blocked by: #23 sub-issue 1 (`adl init`) — config must exist before adding repos

## Testing Requirements

- [ ] Unit tests for config append logic (read → append → write preserves existing entries)
- [ ] Unit tests for scaffolding (copies missing files, skips existing)
- [ ] Unit tests for duplicate repo detection
- [ ] Mock tests for `gh` CLI interaction

### PR DESCRIPTION
**Pull request by rube-de**
## Summary

Closes #25

- Add `adl add [path]` interactive CLI command that onboards a repository by detecting GitHub info via `gh` CLI, mapping project columns, scaffolding agent/workflow files, and appending to `~/.adl/config.yaml`
- Bundle agent markdown and workflow YAML templates inside the package (`auto_dev_loop.bundled`) for runtime scaffolding access
- Auto-detect standard column names (Ready for Dev, In Progress, Done) with case-insensitive matching and fallback to manual selection

### New files
- `src/auto_dev_loop/add_repo.py` — all onboarding logic (GitHub detection, config manipulation, scaffolding, interactive wizard)
- `src/auto_dev_loop/bundled/` — packaged agent/workflow templates
- `tests/test_add_repo.py` — 30 tests covering all helpers and wizard flows

### Modified files
- `src/auto_dev_loop/cli.py` — added `add` command
- `tests/test_cli.py` — 3 new CLI tests

## Test Plan

- [x] 370 tests passing (30 new + 340 existing)
- [ ] Manual: `adl add ~/projects/some-repo` with a real GitHub repo + Projects V2
- [ ] Manual: `adl add` (no path) uses cwd
- [ ] Manual: verify scaffolded files appear in target repo's `agents/` and `workflows/`
- [ ] Manual: verify `adl validate` passes after `adl add`

## Release Notes

* **New Features**
  * Added `adl add` command with interactive wizard to onboard repositories into ADL
  * Wizard detects GitHub remote, lists available projects, and configures status column mapping
  * Added 5 new workflow templates: bug fix, documentation, feature development, ops changes, and security audit
  * Introduced 9 specialized agent roles for planning, development, testing, review, and security workflows
  * GitHub CLI integration with authentication validation and error guidance

### CODE REVIEW

File: src/auto_dev_loop/add_repo.py [hunk 1]
```diff
@@ -0,0 +1,330 @@
...
+    resolved = (repo_path or Path.cwd()).resolve()
+    if not (resolved / ".git").is_dir():
+        typer.echo(f"{resolved} is not a git repository.", err=True)
+        raise typer.Exit(1)
+
+    # 3. Check for duplicates
+    if is_repo_configured(config_path, resolved):
+        typer.echo(f"Repository {resolved} is already configured.")
+        if not typer.confirm("Reconfigure?", default=False):
+            raise typer.Exit(0)
```
[the remark this answers is not in the document]
Comment (rube-de) [reply src/auto_dev_loop/add_repo.py]: Fixed: added `_remove_repo_config` helper and call it on reconfigure before appending, preventing duplicate entries.

File: src/auto_dev_loop/add_repo.py [hunk 2]
```diff
@@ -0,0 +1,330 @@
...
+        ["gh", "project", "list", "--owner", owner, "--format", "json"],
+        capture_output=True,
+        text=True,
+    )
+    if result.returncode != 0:
+        raise AddRepoError(
+            f"Could not list projects for {owner}: {result.stderr.strip()}"
+        )
+    data = json.loads(result.stdout)
+    return data.get("projects", [])
```
[the remark this answers is not in the document]
Comment (rube-de) [reply src/auto_dev_loop/add_repo.py]: Fixed: wrapped both `json.loads` calls (line 129 and 161) in try/except `JSONDecodeError`, raising `AddRepoError` with context.

File: src/auto_dev_loop/add_repo.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (rube-de) [reply src/auto_dev_loop/add_repo.py]: Fixed: added `_remove_repo_config` helper to strip existing entry before appending on reconfigure.

File: src/auto_dev_loop/add_repo.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (rube-de) [reply src/auto_dev_loop/add_repo.py]: Fixed: added try/except `JSONDecodeError` guard around `json.loads` in both `list_gh_projects` and `list_status_options`.

File: src/auto_dev_loop/cli.py [hunk 3]
```diff
@@ -75,3 +75,21 @@ def init(
     from .init_wizard import run_init_wizard
 
     run_init_wizard(config)
+
+
+@app.command()
+def add(
+    path: Path = typer.Argument(
```
[the remark this answers is not in the document]
Comment (rube-de) [reply src/auto_dev_loop/cli.py]: Fixed: annotated `path` parameter as `Path | None` to match runtime behavior.

File: src/auto_dev_loop/add_repo.py [hunk 4]
```diff
@@ -0,0 +1,330 @@
...
+    return False
+
+
+def append_repo_config(config_path: Path, entry: dict[str, Any]) -> None:
+    """Append a repo entry to the config's repos list and write back."""
+    data = load_config_raw(config_path)
+    repos = data.get("repos", [])
+    repos.append(entry)
+    data["repos"] = repos
+    config_path.write_text(yaml.safe_dump(data, sort_keys=False))
```
[the remark this answers is not in the document]
Comment (rube-de) [reply src/auto_dev_loop/add_repo.py]: Fixed: changed to `data.get("repos") or []` and added `isinstance` check for list type.

File: src/auto_dev_loop/bundled/agents/pr_fixer.md [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+- Address each comment specifically
+- Don't introduce unrelated changes
+- Run tests after each fix
+- Push changes to the PR branch
+
+## Output
+
+End your response with:
+- `<<<VERDICT:FIXES_APPLIED>>>` — all comments addressed
+- `BLOCKED` — some comments require human clarification
```
[the remark this answers is not in the document]
Comment (rube-de) [reply src/auto_dev_loop/bundled/agents/pr_fixer.md]: Fixed: updated to `<<<VERDICT:BLOCKED>>>` marker format.

File: src/auto_dev_loop/add_repo.py [hunk 6]
```diff
@@ -0,0 +1,330 @@
...
+        if dest.exists():
+            continue
+        shutil.copy2(src_file, dest)
+        copied.append(src_file.name)
+    return copied
+
+
+def load_config_raw(config_path: Path) -> dict[str, Any]:
+    """Load config YAML as a raw dict (no dataclass parsing)."""
+    return yaml.safe_load(config_path.read_text()) or {}
```
[the remark this answers is not in the document]
Comment (rube-de) [reply src/auto_dev_loop/add_repo.py]: Acknowledged — will be addressed by the author

File: src/auto_dev_loop/bundled/__init__.py [hunk 7]
```diff
@@ -0,0 +1,8 @@
+"""Bundled agent and workflow templates for scaffolding new repos."""
+
+from pathlib import Path
+
+_BUNDLED_ROOT = Path(__file__).parent
```
[the remark this answers is not in the document]
Comment (rube-de) [reply src/auto_dev_loop/bundled/__init__.py]: Acknowledged — will be addressed by the author

File: src/auto_dev_loop/bundled/agents/orchestrator.md [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+4. When developer applies fixes, trigger tester re-run
+5. Continue until tests pass or max iterations reached
+
+## Coordination
+
+- Tester writes results to `.adl-output/test-results.json`
+- Developer reads test results and writes patches
+- Use Bash polling to check for output files
+- Use SendMessage for direct P2P communication when both agents are active
+
```
[the remark this answers is not in the document]
Comment (rube-de) [reply src/auto_dev_loop/bundled/agents/orchestrator.md]: Acknowledged — will be addressed by the author

File: src/auto_dev_loop/add_repo.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (rube-de) [reply src/auto_dev_loop/add_repo.py]: Fixed: added docstring noting FileNotFoundError precondition.

File: src/auto_dev_loop/add_repo.py [hunk 9]
```diff
@@ -0,0 +1,330 @@
...
+        for role, col in defaults.items():
+            typer.echo(f"  {role}: {col}")
+        if typer.confirm("Use these column mappings?", default=True):
+            return defaults
+
+    typer.echo("Map project columns:")
+    columns = {}
+    for role in ("source", "in_progress", "done"):
+        columns[role] = _prompt_column(role, options, defaults.get(role))
+    return columns
```
[the remark this answers is not in the document]
Comment (rube-de) [reply src/auto_dev_loop/add_repo.py]: Fixed: detected columns are now shown as defaults in the prompt, allowing the user to override them.

File: src/auto_dev_loop/bundled/__init__.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (rube-de) [reply src/auto_dev_loop/bundled/__init__.py]: Fixed: switched to `importlib.resources.files()` for resolving bundled template paths.

File: src/auto_dev_loop/bundled/agents/orchestrator.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (rube-de) [reply src/auto_dev_loop/bundled/agents/orchestrator.md]: Acknowledged — tracked in #35

**Comment by rube-de** [comment]
## PR Comment Status

| Status | Threads | Review Bodies | Total |
|--------|---------|---------------|-------|
| Resolved | 0 | 3 | 3 |
| Fixed by DLC | 21 | 0 | 21 |
| Skipped (user decision) | 3 | 0 | 3 |
| Discussion (needs human) | 0 | 0 | 0 |
| Blocked | 0 | 0 | 0 |
| Dismissed | 0 | 0 | 0 |
| Deferred to issue | 2 | 0 | 2 |
| **Total** | **26** | **3** | **29** |

## Decisions

- `add_repo.py:299` — Skipped: repetitive try/except is acceptable (4 blocks, minimal overhead)
- `add_repo.py:64` — Skipped: yaml.safe_dump comment-stripping is a cross-cutting concern out of scope
- `orchestrator.md:26` — Deferred: template/runtime mismatch tracked in #35
- `bug_fix.yaml:28` — Deferred: per-stage reviewers not honored by engine, tracked in #35

## Follow-up

Follow-up issue: #35

Author will address some remaining items manually.

File: src/auto_dev_loop/add_repo.py [hunk 10]
```diff
@@ -0,0 +1,388 @@
...
+    if not projects:
+        typer.echo(
+            f"No GitHub Projects V2 found for {owner}.\n"
+            f"Create one at https://github.com/orgs/{owner}/projects "
+            f"or https://github.com/users/{owner}/projects",
+            err=True,
+        )
+        raise typer.Exit(1)
+
+    project = _prompt_project(projects)
```
[the remark this answers is not in the document]
Comment (rube-de) [reply src/auto_dev_loop/add_repo.py]: Fixed: wrapped `_prompt_project()` call in try/except AddRepoError, consistent with all other error-raising calls in the wizard.

File: src/auto_dev_loop/add_repo.py [hunk 11]
```diff
@@ -0,0 +1,411 @@
...
+    data = load_config_raw(config_path)
+    repos = data.get("repos") or []
+    if not isinstance(repos, list):
+        raise AddRepoError(
+            f"Config file {config_path} is invalid: 'repos' must be a list."
+        )
+    repos.append(entry)
+    data["repos"] = repos
+    config_path.write_text(yaml.safe_dump(data, sort_keys=False))
+
```
[the remark this answers is not in the document]
Comment (rube-de) [reply src/auto_dev_loop/add_repo.py]: Fixed: added type guard for null/non-string path entries in is_repo_configured.

File: src/auto_dev_loop/add_repo.py [hunk 12]
```diff
@@ -0,0 +1,411 @@
...
+    target_dir.mkdir(parents=True, exist_ok=True)
+    copied: list[str] = []
+    for src_file in sorted(source_dir.iterdir()):
+        if not src_file.is_file():
+            continue
+        dest = target_dir / src_file.name
+        if dest.exists():
+            continue
+        shutil.copy2(src_file, dest)
+        copied.append(src_file.name)
```
[the remark this answers is not in the document]
Comment (rube-de) [reply src/auto_dev_loop/add_repo.py]: Fixed: replaced shutil.copy2 with read_bytes/write_bytes for Traversable compatibility.

File: src/auto_dev_loop/add_repo.py [hunk 13]
```diff
@@ -0,0 +1,411 @@
...
+    result = subprocess.run(
+        ["gh", "auth", "status"],
+        capture_output=True,
+        text=True,
+        timeout=10,
+    )
+    if result.returncode != 0:
+        raise AddRepoError(
+            "GitHub CLI is not authenticated. Run `gh auth login` first."
+        )
```
[the remark this answers is not in the document]
Comment (rube-de) [reply src/auto_dev_loop/add_repo.py]: Fixed: wrapped gh auth status call in try/except TimeoutExpired, consistent with the first subprocess call.

File: src/auto_dev_loop/add_repo.py [hunk 14]
```diff
@@ -0,0 +1,411 @@
...
+        columns = {
+            "source": typer.prompt(
+                "Column name for items ready for development",
+                default="Ready for Dev",
+            ),
+            "in_progress": typer.prompt(
+                "Column name for items in progress",
+                default="In Progress",
+            ),
+            "done": typer.prompt(
```
[the remark this answers is not in the document]
Comment (rube-de) [reply src/auto_dev_loop/add_repo.py]: Fixed: added 'repo' field to config entry and RepoConfig dataclass.

File: src/auto_dev_loop/add_repo.py [hunk 15]
```diff
@@ -0,0 +1,411 @@
...
+                "Column name for completed items",
+                default="Done",
+            ),
+        }
```
[the remark this answers is not in the document]
Comment (rube-de) [reply src/auto_dev_loop/add_repo.py]: Fixed: added uniqueness loop in the no-status-field fallback, matching the pattern in _prompt_columns.

File: tests/test_add_repo.py [hunk 16]
```diff
@@ -0,0 +1,441 @@
...
+        data["repos"] = [{"path": str(repo_path), "project_number": 1}]
+        _write_config(cfg_path, data)
+
+        monkeypatch.setattr(
+            "auto_dev_loop.add_repo.typer.confirm", lambda *a, **kw: False
+        )
+
+        with pytest.raises(typer.Exit) as exc_info:
+            run_add_wizard(repo_path, cfg_path)
+        assert exc_info.value.exit_code == 0
```
[the remark this answers is not in the document]
Comment (rube-de) [reply tests/test_add_repo.py]: Fixed: added test_reconfigure_replaces_existing_entry and test_no_status_field_prompts_manual_columns tests.

File: src/auto_dev_loop/models.py [hunk 17]
```diff
@@ -14,6 +14,13 @@
 VERDICT_IMPLEMENTATION_COMPLETE = "<<<VERDICT:IMPLEMENTATION_COMPLETE>>>"
 VERDICT_FIXES_APPLIED = "<<<VERDICT:FIXES_APPLIED>>>"
 VERDICT_FEEDBACK_APPLIED = "<<<VERDICT:FEEDBACK_APPLIED>>>"
+# TODO(#35): These verdict constants are defined for agent prompt consistency
+# but are not yet handled by _parse_verdict() in workflow_engine.py.
+# They will silently fall through to the default verdict ("approved" in
+# non-strict mode) until #35 adds proper state machine support.
+VERDICT_BLOCKED = "<<<VERDICT:BLOCKED>>>"
+VERDICT_CLARIFICATION_NEEDED = "<<<VERDICT:CLARIFICATION_NEEDED>>>"
+VERDICT_MAX_ITERATIONS = "<<<VERDICT:MAX_ITERATIONS>>>"
```
[the remark this answers is not in the document]
Comment (rube-de) [reply src/auto_dev_loop/models.py]: Acknowledged — not applicable. The config YAML is machine-managed, not hand-edited. Same concern already addressed in earlier thread.

File: src/auto_dev_loop/add_repo.py [hunk 18]
```diff
@@ -0,0 +1,411 @@
...
+def is_repo_configured(config_path: Path, repo_path: Path) -> bool:
+    """Check if a repo path is already in the config."""
+    data = load_config_raw(config_path)
+    target = str(repo_path.resolve())
+    for entry in data.get("repos", []):
+        if not isinstance(entry, dict):
+            continue
+        existing = entry.get("path", "")
+        if str(Path(existing).expanduser().resolve()) == target:
+            return True
```
[the remark this answers is not in the document]
Comment (rube-de) [reply src/auto_dev_loop/add_repo.py]: Fixed: added type guard for null/non-string path entries — skips entries with missing or null path values.

File: src/auto_dev_loop/add_repo.py [hunk 19]
```diff
@@ -0,0 +1,446 @@
...
+        typer.echo(f"Scaffolded workflows: {', '.join(workflows_copied)}")
+    if not agents_copied and not workflows_copied:
+        typer.echo("All agent/workflow files already exist, nothing scaffolded.")
+
+    # 9. Append to config
+    entry = {
+        "path": str(resolved),
+        "project_number": project["number"],
+        "owner": owner,
+        "repo": repo,
```
[the remark this answers is not in the document]
Comment (rube-de) [reply src/auto_dev_loop/add_repo.py]: Fixed: added `r.get("repo")` to config.py kwargs so the repo field is now parsed into RepoConfig.

File: src/auto_dev_loop/bundled/workflows/security_audit.yaml [hunk 20]
```diff
@@ -0,0 +1,12 @@
+id: security_audit
+description: Security audit — security-first with veto authority
+
+stages:
+  - ref: security
+    agent: security_reviewer
+    canVeto: true
+
+  - ref: code_review
+    agent: reviewer
```
[the remark this answers is not in the document]
Comment (rube-de) [reply src/auto_dev_loop/bundled/workflows/security_audit.yaml]: Fixed: added `reviewers: [reviewer]` to enable strict verdict parsing for the code_review stage.

File: src/auto_dev_loop/add_repo.py [hunk 21]
```diff
@@ -0,0 +1,446 @@
...
+                    default="Done",
+                ),
+            }
+            if len(set(columns.values())) == 3:
+                break
+            typer.echo("  Each role must map to a different column. Please try again.")
+
+    # 8. Scaffold agents and workflows
+    agents_copied = scaffold_files(BUNDLED_AGENTS_DIR, resolved / "agents")
+    workflows_copied = scaffold_files(BUNDLED_WORKFLOWS_DIR, resolved / "workflows")
```
[the remark this answers is not in the document]
Comment (rube-de) [reply src/auto_dev_loop/add_repo.py]: Acknowledged — tracked in #37

**Comment by rube-de** [comment]
## PR Comment Status

| Status | Threads | Review Bodies | Total |
|--------|---------|---------------|-------|
| Resolved | 47 | 4 | 51 |
| Fixed by DLC | 14 | 0 | 14 |
| Skipped (user decision) | 0 | 0 | 0 |
| Discussion (needs human) | 1 | 0 | 1 |
| Blocked | 0 | 0 | 0 |
| Dismissed | 0 | 0 | 0 |
| **Total** | **62** | **4** | **66** |

## Decisions

- `src/auto_dev_loop/add_repo.py:30` — Fixed: sorted() key for Traversable compatibility
- `src/auto_dev_loop/add_repo.py:88` — Fixed: _remove_repo_config type validation
- `src/auto_dev_loop/add_repo.py:226` — Fixed: opt.get("name") defensive key access
- `src/auto_dev_loop/add_repo.py:338` — Fixed: .exists() already handles worktrees/submodules
- `src/auto_dev_loop/add_repo.py:443` — Fixed: added max_attempts guard to no-status-field loop
- `src/auto_dev_loop/add_repo.py:460` — Fixed: repo field now parsed by config.py
- `src/auto_dev_loop/add_repo.py:447` — Discussion: per-repo agent/workflow paths deferred to #37
- `src/auto_dev_loop/bundled/agents/security_reviewer.md:31` — Fixed: verdict marker instructions require own line
- `src/auto_dev_loop/bundled/agents/tester.md:22` — Fixed: TESTS_FAILING → VERDICT:NEEDS_REVISION
- `src/auto_dev_loop/bundled/workflows/security_audit.yaml:10` — Fixed: added reviewers for strict parsing
- `src/auto_dev_loop/config.py:82` — Fixed: repo field now extracted in config loader
- `tests/test_add_repo.py:37` — Fixed: glob to iterdir for Traversable compatibility (agents + workflows)
- `tests/test_add_repo.py:366` — Fixed: added repo field assertion to happy path + reconfigure tests

## Follow-up

Follow-up issue: #37

File: src/auto_dev_loop/add_repo.py [hunk 22]
```diff
@@ -0,0 +1,477 @@
...
+    """Append a repo entry to the config's repos list and write back."""
+    data = load_config_raw(config_path)
+    repos = data.get("repos") or []
+    if not isinstance(repos, list):
+        raise AddRepoError(
+            f"Config file {config_path} is invalid: 'repos' must be a list."
+        )
+    repos.append(entry)
+    data["repos"] = repos
+    config_path.write_text(yaml.safe_dump(data, sort_keys=False))
```
[the remark this answers is not in the document]
Comment (rube-de) [reply src/auto_dev_loop/add_repo.py]: Acknowledged — will be addressed by the author

[the remark this answers is not in the document]
Comment (rube-de) [reply src/auto_dev_loop/add_repo.py]: Fixed: switched config write operations to ruamel.yaml round-trip mode to preserve comments and formatting.

File: src/auto_dev_loop/add_repo.py [hunk 23]
```diff
@@ -0,0 +1,518 @@
...
+def _load_config_rt(config_path: Path) -> Any:
+    """Load config via round-trip YAML, returning an empty dict for empty files."""
+    data = _RT_YAML.load(config_path)
+    if data is None:
+        return {}
+    if not isinstance(data, dict):
+        raise AddRepoError(
+            f"Config file {config_path} must contain a YAML mapping at the top level."
+        )
+    return data
```
[the remark this answers is not in the document]
Comment (funny-otter) [reply src/auto_dev_loop/add_repo.py]: Fixed: wrapped `_RT_YAML.load()` in `try/except YAMLError`, raising `AddRepoError` with context — consistent with `load_config_raw`.
